impl/src/expand.rs

```diff
@@ -377,6 +547,132 @@ fn impl_enum(input: Enum) -> TokenStream {
     }
 }
 
+#[cfg_attr(test, derive(Debug))]
+#[derive(Clone)]
+struct GenericUsageVisitor {
+    generics: std::collections::HashMap<proc_macro2::Ident, bool>,
+}
+
+impl GenericUsageVisitor {
+    pub fn new<TPairs>(generics: TPairs) -> Self
+    where
+        TPairs: IntoIterator<Item = (syn::Ident, bool)>,
+    {
+        Self {
+            generics: generics.into_iter().collect(),
+        }
+    }
+
+    pub fn new_unmarked<TIdents>(generics: TIdents) -> Self
+    where
+        TIdents: IntoIterator<Item = syn::Ident>,
+    {
+        Self::new(generics.into_iter().map(|ident| (ident, false)))
+    }
+
+    pub fn iter_marked(&self) -> impl Iterator<Item = &proc_macro2::Ident> {
+        self.generics.iter().filter(|(_k, v)| **v).map(|(k, _v)| k)
+    }
+
+    #[allow(dead_code)]
+    pub fn is_marked<'a, T: PartialEq<&'a proc_macro2::Ident> + 'a>(&'a self, item: T) -> bool {
+        self.iter_marked().any(|ident| item == ident)
+    }
+
+    pub fn mark_from<'a, TMarkedSource: IntoIterator<Item = &'a proc_macro2::Ident> + 'a>(
+        &'a mut self,
+        source: TMarkedSource,
+    ) -> bool {
+        let mut any_marked = false;
+        for other_marked in source {
+            if let Some(is_marked) = self.generics.get_mut(other_marked) {
+                *is_marked = true;
+                any_marked = true;
+            }
+        }
+        any_marked
+    }
+}
+
+impl<'ast> syn::visit::Visit<'ast> for GenericUsageVisitor {
+    fn visit_type_path(&mut self, i: &'ast syn::TypePath) {
+        if let Some(ident) = i.path.get_ident() {
+            if let Some(entry) = self.generics.get_mut(ident) {
+                *entry = true;
+            }
+        }
+        syn::visit::visit_type_path(self, i);
+    }
+
+    fn visit_type_param(&mut self, i: &'ast syn::TypeParam) {
```

Comment:
This is not reachable &mdash; except maybe in contrived code like `field: [T; { fn f<T>() {}; 0 }]` where you actually don't want the behavior here of considering the inner `T` TypeParam as being a use of the outer `T`.

Reply:
I used this to deep-search for generic usages because- when a generic wasn't used- I was running into circumstances where fixed types would succeed in compilation where they should not, because it generated impossible/unlikely bounds, such as `PathBuf: Display`, which you seem to have resolved using the `PathAsDisplay` trait in the past. My solution was to deep-scan for generics in types, and only add bounds for fields which used at least one generic.